platform-hub-web/src/app/app.routes.spec.js

```diff
@@ -0,0 +1,373 @@
+  let Me = null;
+
+  beforeEach(() => {
+    const moduleName = 'app.routes.spec';
+    angular.module(moduleName, ['app']);
+    angular.mock.module(moduleName);
+  });
+
+  beforeEach(angular.mock.inject((_$q_, _$httpBackend_, _$state_, _$location_, _$rootScope_, _authService_, _loginDialogService_, _FeatureFlags_, _roleCheckerService_, _Me_) => {
+    sandbox = sinon.sandbox.create();
+
+    $q = _$q_;
+    $httpBackend = _$httpBackend_;
+    $state = _$state_;
+    $location = _$location_;
+    $rootScope = _$rootScope_;
+    authService = _authService_;
+    loginDialogService = _loginDialogService_;
+    FeatureFlags = _FeatureFlags_;
+    roleCheckerService = _roleCheckerService_;
+    Me = _Me_;
+
+    $httpBackend
+      .whenGET('/api/app_settings')
+      .respond('{}');
+  }));
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  describe('route config', () => {
+    it(`should only allow certain fields in the 'data' object`, () => {
+      function stubs() { }
+
+      function checks(state) {
+        lodash.keys(state.data).forEach(k => {
+          expect(ALLOWED_DATA_CONFIG_FIELDS).toContain(k, `in 'data' config for route: ${state.name}`);
+        });
+      }
+
+      eachNavigableState(stubs, checks);
+    });
+  });
+
+  describe(`for not logged in users who don't log in`, () => {
+    it(`should only transition to states that don't need authentication`, () => {
+      console.log(`\n*** Not logged in users who don't log in`);
+
+      function stubs() {
+        sandbox.stub(authService, 'isAuthenticated').returns(false);
+
+        sandbox.stub(loginDialogService, 'run').usingPromise($q).rejects();
+
+        sandbox.spy(FeatureFlags, 'isEnabled');
+
+        sandbox.spy(roleCheckerService, 'hasHubRole');
+      }
+
+      function checks(state) {
+        if (state.data.authenticate) {
+          authService.isAuthenticated.should.have.callCount(1);
+          loginDialogService.run.should.have.callCount(1);
+          FeatureFlags.isEnabled.should.have.callCount(0);
+          roleCheckerService.hasHubRole.should.have.callCount(0);
+          expect($state.current.name).toBe('home');
+        } else {
+          authService.isAuthenticated.should.have.callCount(0);
+          loginDialogService.run.should.have.callCount(0);
+          FeatureFlags.isEnabled.should.have.callCount(0);
+          roleCheckerService.hasHubRole.should.have.callCount(0);
+          expect($state.current.name).toBe(state.name);
+        }
+      }
+
+      eachNavigableState(stubs, checks);
+    });
+  });
+
+  describe('for not logged in users who log in', () => {
+
+    // We currently assume that only the 'admin' role is used.
+    // If new roles are added later, these tests will need to be updated, and
+    // rethought (in terms of structure).
+
+    describe('when all feature flags are turned on', () => {
+
+      describe('for non-admin users', () => {
+        it(`should transition to states as expected`, () => {
```

Comment:
It's really hard to describe all the possible permutations for the states – any suggestions on what this could say?